### PR TITLE
[GVN][NewGVN] Take call attributes into account in expressions

### PR DESCRIPTION
**Pull request by dtcxzyw**
Drop `canBeReplacedBy` and take call attributes into account in expressions.
Address comment https://github.com/llvm/llvm-project/pull/114011#pullrequestreview-2409772313.

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/GVN.cpp [hunk 1]
```diff
@@ -143,6 +143,8 @@ struct llvm::GVNPass::Expression {
   Type *type = nullptr;
   SmallVector<uint32_t, 4> varargs;
 
+  std::optional<AttributeList> attrs;
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/GVN.cpp:146]: You don't need std::optional, AttributeList is its own optional.

**Comment by dtcxzyw** [comment]
Compile-time impact: http://llvm-compile-time-tracker.com/compare.php?from=155fed8017812fe451720f79d34c8a7287c9a24d&to=a84bc7a133d7f56047acdf4230470ce4caf675f5&stat=instructions%3Au

**Comment by dtcxzyw** [comment]
Ping.

**Review by nikic — APPROVED** [review]
LGTM. This is ugly but I don't have a better idea.

**Comment by ronlieb** [comment]
getting traceback, see attached 
[traceback.txt](https://github.com/user-attachments/files/17685200/traceback.txt)

**Comment by dtcxzyw** [comment]
> getting traceback, see attached [traceback.txt](https://github.com/user-attachments/files/17685200/traceback.txt)

Can you please provide a reproducer?

**Comment by ronlieb** [comment]
yes, but it will be a day or two

**Pull request merged**